File: donghwa/arxiv/MST/1584_prim.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Define a structure to represent points
struct Point {
    int x, y;
};

// Function to calculate the Manhattan distance between two points
int manhattanDist(struct Point p1, struct Point p2) {
    return abs(p1.x - p2.x) + abs(p1.y - p2.y);
}

int minCostConnectPoints(int** points, int pointsSize, int* pointsColSize) {
    int n = pointsSize;
    if (n <= 1) {
        return 0;
    }

    struct Point* vertices = (struct Point*)malloc(n * sizeof(struct Point));
    for (int i = 0; i < n; i++) {
        vertices[i].x = points[i][0];
        vertices[i].y = points[i][1];
    }

    int* minHeap = (int*)malloc(n * sizeof(int));
    int* costs = (int*)malloc(n * sizeof(int));
    bool* visited = (bool*)malloc(n * sizeof(bool));

    // Initialize data structures
    for (int i = 0; i < n; i++) {
        minHeap[i] = i;
        costs[i] = INT_MAX;
        visited[i] = false;
    }

    costs[0] = 0;
    int minCost = 0;

    while (1) {
        int minCostVertex = -1;
        for (int i = 0; i < n; i++) {
            if (!visited[i] && (minCostVertex == -1 || costs[i] < costs[minCostVertex])) {
                minCostVertex = i;
            }
        }

        if (minCostVertex == -1) {
            break; // All vertices are visited
        }

        visited[minCostVertex] = true;
        minCost += costs[minCostVertex];

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {

File: donghwa/arxiv/MST/1584_prim.py

import heapq

def minCostConnectPoints(points):
    n = len(points)
    if n <= 1:
        return 0

    def manhattan_dist(p1, p2):
        return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])

    visited = [False] * n
    min_heap = [(0, 0)]  # (cost, vertex)
    min_cost = 0

    while min_heap:
        cost, vertex = heapq.heappop(min_heap)
        if visited[vertex]:
            continue

        visited[vertex] = True
        min_cost += cost

        for i in range(n):
            if not visited[i]:
                heapq.heappush(min_heap, (manhattan_dist(points[vertex], points[i]), i))

    return min_cost

# Test the function with the given example
points = [[0,0],[2,2],[3,10],[5,2],[7,0]]
print(minCostConnectPoints(points))  # Output should be 20
